src/Unsafe/Linear.hs

```diff
@@ -20,42 +27,188 @@
+-- 'toLinear3' = toLinearN \@3
+-- @
+toLinearN :: forall n f g m. ToLinearN n f g => f %m-> g
+toLinearN f = case unsafeLinearityProofN @n @f @g of
+  UnsafeRefl -> f
+
+-- | @ToLinearN n f g@ means that @f@ and @g@ are the same with the
+-- possible exception of the multiplicities of the first @n@ arrows.
+type ToLinearN :: forall {rep :: RuntimeRep}. Nat -> TYPE rep -> TYPE rep -> Constraint
+class ToLinearN n f g where
+  -- | @unsafeLinearityProofN \@n \@f \@g @ is a fake proof that @f@ and @g@ are
+  -- the same with the possible exception of the multiplicities of the first
+  -- @n@ arrows.
+  unsafeLinearityProofN :: UnsafeEquality f g
+instance (ToLinearN' ni f g, UnifyL ni f g, UnifyL ni g f, ni ~ ToINat n) => ToLinearN n f g where
+  unsafeLinearityProofN = prf @ni
+
+-- | @toLinearL@ is similar to 'toLinearN', but it offers
+-- finer-grained control. The user supplies a type argument
+-- showing what they expect each of the arrows to be originally,
+-- and what each changes to. These may be polymorphic, and wildcards
+-- will work when they can be inferred.
+--
+-- === Examples
+--
+-- @
+-- toLinearL \@'[ 'Many :=> 'One, 'Many :=> 'One]
+--   :: (a -> b -> c) -> a %1-> b %1-> c
+-- 'toLinear3' = toLinearL \@'[ _ :=> 'One, _ :=> 'One, _ :=> 'One]
+-- @
+toLinearL :: forall (l :: [Change]) f g m. ToLinearL l f g => f %m-> g
+toLinearL f = case unsafeLinearityProofL @l @f @g of
+  UnsafeRefl -> f
+
+-- | @ToLinearL l f g@ means that @f@ and @g@ are the same with the
+-- possible exception of the multiplicities of some of the first arrows,
+-- as described by the list @l@.
+type ToLinearL :: forall {rep :: RuntimeRep}. [Change] -> TYPE rep -> TYPE rep -> Constraint
+class ToLinearL l f g where
+  -- | @unsafeLinearityProofL \@l \@f \@g @ is a fake proof that @f@ and @g@ are
+  -- the same with the possible exception of the multiplicities of the first
+  -- arrows given by the list @l@.
+  unsafeLinearityProofL :: UnsafeEquality f g
+instance ( ToLinearN' ni f g, UnifyL ni f g, UnifyL ni g f
+         , ni ~ Length l, UseChanges l f g) => ToLinearL l f g where
+  unsafeLinearityProofL = prf @ni
+
+-- | Convert a GHC 'Nat' to a real inductive natural number.
+-- We use this because GHC 'Nat' offers a friendly API but
+-- it's a terrible pain for implementation.
+type ToINat :: Nat -> INat
+type family ToINat n where
+  ToINat 0 = 'Z
+  ToINat n = 'S (ToINat (n - 1))
+
+-- | The actual implementation of 'ToLinearN', using the inductive natural
+-- number it's handed.
+type ToLinearN' :: forall {rep :: RuntimeRep}. INat -> TYPE rep -> TYPE rep -> Constraint
+class ToLinearN' arrs f g where
+  prf :: UnsafeEquality f g
+instance ToLinearN' k fb gb => ToLinearN' ('S k) ((a :: TYPE repa) %p-> (fb :: TYPE repb)) (a %q-> (gb :: TYPE repb)) where
+  prf = case prf @k @fb @gb of
+    UnsafeRefl -> case unsafeEqualityProof @p @q of
+      UnsafeRefl -> UnsafeRefl
+instance ToLinearN' 'Z (a :: TYPE rep) (a :: TYPE rep) where
+  prf = UnsafeRefl
```

Comment:
Does this produce an O(n) irreducible proof? I'm not super up to date with the new `unsafeEqualityProof` mechanism.

And if so: is it bad?

Reply:
The proof can get a bit long, but I don't think that we can do anything about it. It'll be erased in Core Prep or so.